src/Juvix/Core/IR/Subst.hs

```diff
@@ -0,0 +1,105 @@
+-- |
+-- - Runs a substitution algorithm over core
+module Juvix.Core.IR.Subst where
+
+import Control.Lens
+import qualified Data.HashSet as Set
+import qualified Juvix.Core.Common.NameSymbol as NameSymbol
+import qualified Juvix.Core.IR.Types as Types
+import qualified Juvix.Core.Usage as Usage
+import Juvix.Library
+import qualified Juvix.Library.HashMap as Map
+
+-- eventually a lot of this code will be replaced by a proper
+-- inliner strategy, but this code will serve as the basis for
+-- said iterative design
+
+data Key
+  = Bound Natural
+  | Global NameSymbol.T
+  deriving (Show, Eq, Generic)
+
+instance Hashable Key
+
+data T primTy primVal
+  = T
+      { _depth :: Natural,
+        _sub :: Map.T Key (Types.Elim primTy primVal)
+      }
+  deriving (Show)
+
+makeLenses ''T
```

Comment:
alright... are they really worth it? relatively simple transformations

(I don't mind lenses but we only use them sometimes, it's a bit inconsistent)

Reply:
I'm using it for over, which is useful, nicer than using the record updating syntax for a single field.